tapiriik/services/Hexoskin/hexoskin.py

```diff
@@ -0,0 +1,256 @@
+            "/api/activitytype/3/": ActivityType.CrossCountrySkiing,
+            "/api/activitytype/4/": ActivityType.DownhillSkiing,
+            "/api/activitytype/5/": ActivityType.Hiking,
+            "/api/activitytype/6/": ActivityType.Running,
+            "/api/activitytype/7/": ActivityType.MountainBiking,
+            "/api/activitytype/9/": ActivityType.Rowing,
+            "/api/activitytype/10/": ActivityType.Running,
+            "/api/activitytype/11/": ActivityType.Skating,
+            "/api/activitytype/13/": ActivityType.Snowboarding,
+            "/api/activitytype/14/": ActivityType.Swimming,
+            "/api/activitytype/15/": ActivityType.Walking,
+            "/api/activitytype/24/": ActivityType.Running,
+            "/api/activitytype/28/": ActivityType.Gym,
+        }
+        if key in _reverseActivityTypeMappingsKeys.keys():
+            return _reverseActivityTypeMappingsKeys[key]
+        else:
+            return ActivityType.Other
+
+    SupportedActivities = list(_activityTypeMappings.keys())
+
+
+    def UserUploadedActivityURL(self, uploadId):
+        return serverRoot + "api/range/%d/" % uploadId
+
+
+    def WebInit(self):
+        """
+        prepare the oauth process request. Done separately because it needs to be
+        initialized on page display
+        """
+        from uuid import uuid4
+        params = {'scope':'readwrite',
+                  'client_id':HEXOSKIN_CLIENT_ID,
+                  'response_type':'code',
+                  'state': str(uuid4()),
+                  'redirect_uri':WEB_ROOT + reverse("oauth_return", kwargs={"service": "hexoskin"})}
+        self.UserAuthorizationURL = serverRoot + "api/connect/oauth2/auth/?" + urlencode(params)
+
+
+    def _apiHeaders(self, serviceRecord):
+        return {"Authorization": "Bearer " + serviceRecord.Authorization["OAuthToken"]}
+
+
+    def RetrieveAuthorizationToken(self, req, level):
+        """In OAuth flow, retrieve the Authorization Token"""
+        code = req.GET.get("code")
+        params = {"grant_type": "authorization_code", "code": code,"client_id":HEXOSKIN_CLIENT_ID, "client_secret": HEXOSKIN_CLIENT_SECRET, "redirect_uri": WEB_ROOT + reverse("oauth_return", kwargs={"service": "hexoskin"})}
+        path = serverRoot + "api/connect/oauth2/token/"
+        response = requests.post(path, params=params, auth=(HEXOSKIN_CLIENT_ID,HEXOSKIN_CLIENT_SECRET))
+
+        if response.status_code != 200:
+            raise APIException("Invalid code")
+
+        data = response.json()
+        authorizationData = {"OAuthToken": data["access_token"]}
+        id_resp = requests.get(serverRoot + "api/account/", headers=self._apiHeaders(ServiceRecord({"Authorization": authorizationData})))
+        return (id_resp.json()['objects'][0]['id'], authorizationData)
+
+
+    def RevokeAuthorization(self, serviceRecord):
+        """Delete authorization token"""
+        path = serverRoot + "api/oauth2token/232/"

```

Comment:
Sorry, that slipped my mind when sending that. OAuth tokens are accessed through ids on Hexoskin, so revoke calls must be sent to the token id. Because the token is accessed through it's ID rather than using the public or private key as many other do, i hardcoded mine in the code. The /232/ string would probably be better as an entry in the local_settings though (something like "HEXOSKIN_CLIENT_KEY_ID=XXX")